gexec: Start panics with non-pointer receiver implementation of io.Writer

# My setup
```
go version go1.9 darwin/amd64
```
Gomega version - `dcabb60a477c2b6f456df65037cb6708210fbb02`

# What I did
I run the program bellow (link to playground - https://play.golang.org/p/AF6vrN9IWs).

```go
package main

import (
	"os/exec"

	"github.com/onsi/gomega/gexec"
)

func main() {
	var d discard
	gexec.Start(exec.Command("whoami"), d, d)
}

type discard int

func (d discard) Write(p []byte) (int, error) {
	return len(p), nil
}
```

# What I expected to happen
The program exits cleanly.
# What actually happens
Panic.
```
panic: reflect: call of reflect.Value.IsNil on int Value

goroutine 1 [running]:
reflect.Value.IsNil(0x1101ba0, 0x129ee60, 0x82, 0x129ee60)
	/usr/local/go/src/reflect/value.go:995 +0x12e
github.com/onsi/gomega/gexec.Start(0xc4200bc160, 0x11d16a0, 0x129ee60, 0x11d16a0, 0x129ee60, 0x0, 0x0, 0x0)
	/Users/ivan/workspace/Go/src/github.com/onsi/gomega/gexec/session.go:81 +0x21a
main.main()
	/tmp/gexec/main.go:11 +0xef
exit status
```

The panic is caused by the following lines[0], since calling `IsNil` on non-pointers does not make sense.
```go
	if outWriter != nil && !reflect.ValueOf(outWriter).IsNil() {
		commandOut = io.MultiWriter(commandOut, outWriter)
	}

	if errWriter != nil && !reflect.ValueOf(errWriter).IsNil() {
		commandErr = io.MultiWriter(commandErr, errWriter)
	}
```

I can look into submitting a PR if you're ok with that.
Thanks,
Ivan


[0] https://github.com/onsi/gomega/blob/master/gexec/session.go#L81

Changing the code above to
```go
	if outWriter != nil {
		switch reflect.ValueOf(outWriter).Kind() {
		case reflect.Ptr:
			if !reflect.ValueOf(outWriter).IsNil() {
				commandOut = io.MultiWriter(commandOut, outWriter)
			}
		default:
			commandOut = io.MultiWriter(commandOut, outWriter)
		}
	}

	if errWriter != nil {
		switch reflect.ValueOf(errWriter).Kind() {
		case reflect.Ptr:
			if !reflect.ValueOf(errWriter).IsNil() {
				commandErr = io.MultiWriter(commandErr, errWriter)
			}
		default:
			commandErr = io.MultiWriter(commandErr, errWriter)
		}
```

fixes the issue. However, I'm not sure what use-case is targeted by the second part of the check. It guards against the following:

```
type writer struct{}

func (w *writer) Write(p []byte) (int, error) {
	if w == nil {
		panic("write on nil receiver")
	}
	return len(p), nil
}

func concreteWriter() *writer {
	return nil
}

func interfaceWriter() io.Writer {
	return concreteWriter()
}

func main() {
	w := interfaceWriter()
	_, err := gexec.Start(exec.Command("whoami"), w, w)
	if err != nil {
		log.Fatal(err)
	}
	time.Sleep(time.Second)
}
```
Output:
```
panic: write on nil receiver
```

But even this is strange, since it's completely sane to have a pointer receiver that could work even if the pointer is nil, e.g.
```go
type writer struct{}

func (w *writer) Write(p []byte) (int, error) {
	return len(p), nil
}
```

So I suggest that `if outWriter != nil` is sufficient as a check.

I wanted to discard all output sent to `stderr` when I hit this bug, e.g.
```go
gexec.Start(exec.Command("whoami"), os.Stdout, ioutil.Discard)
```
The code above panics with
```
panic: reflect: call of reflect.Value.IsNil on int Value

goroutine 1 [running]:
reflect.Value.IsNil(0x1107680, 0x129f5b8, 0x82, 0x129f5b8)
	/usr/local/go/src/reflect/value.go:995 +0x12e
github.com/onsi/gomega/gexec.Start(0xc4200bc160, 0x11d2460, 0xc42000c018, 0x11d26a0, 0x129f5b8, 0x0, 0x0, 0x0)
	/Users/ivan/workspace/Go/src/github.com/onsi/gomega/gexec/session.go:85 +0x292
main.main()
	/tmp/gexec/main.go:13 +0x93
exit status 2
```

Cheers,
Ivan

--- a/gexec/session.go
+++ b/gexec/session.go
@@ -7,7 +7,6 @@ import (
 	"io"
 	"os"
 	"os/exec"
-	"reflect"
 	"sync"
 	"syscall"
 
@@ -78,11 +77,11 @@ func Start(command *exec.Cmd, outWriter io.Writer, errWriter io.Writer) (*Sessio
 
 	commandOut, commandErr = session.Out, session.Err
 
-	if outWriter != nil && !reflect.ValueOf(outWriter).IsNil() {
+	if outWriter != nil {
 		commandOut = io.MultiWriter(commandOut, outWriter)
 	}
 
-	if errWriter != nil && !reflect.ValueOf(errWriter).IsNil() {
+	if errWriter != nil {
 		commandErr = io.MultiWriter(commandErr, errWriter)
 	}
 

--- a/gexec/session_test.go
+++ b/gexec/session_test.go
@@ -2,6 +2,7 @@ package gexec_test
 
 import (
 	"io"
+	"io/ioutil"
 	"os/exec"
 	"syscall"
 	"time"
@@ -324,7 +325,10 @@ var _ = Describe("Session", func() {
 	})
 
 	Context("when wrapping out and err", func() {
-		var outWriterBuffer, errWriterBuffer *Buffer
+		var (
+			outWriterBuffer, errWriterBuffer *Buffer
+		)
+
 		BeforeEach(func() {
 			outWriterBuffer = NewBuffer()
 			outWriter = outWriterBuffer
@@ -343,6 +347,17 @@ var _ = Describe("Session", func() {
 
 			Ω(outWriterBuffer.Contents()).Should(Equal(session.Out.Contents()))
 			Ω(errWriterBuffer.Contents()).Should(Equal(session.Err.Contents()))
+		})
+
+		Context("when discarding the output of the command", func() {
+			BeforeEach(func() {
+				outWriter = ioutil.Discard
+				errWriter = ioutil.Discard
+			})
+
+			It("executes succesfuly", func() {
+				Eventually(session).Should(Exit())
+			})
 		})
 	})
 

--- a/gexec/session_test.go
+++ b/gexec/session_test.go
@@ -1,6 +1,7 @@
 package gexec_test
 
 import (
+	"io"
 	"os/exec"
 	"syscall"
 	"time"
@@ -16,7 +17,7 @@ var _ = Describe("Session", func() {
 	var command *exec.Cmd
 	var session *Session
 
-	var outWriter, errWriter *Buffer
+	var outWriter, errWriter io.Writer
 
 	BeforeEach(func() {
 		outWriter = nil
@@ -323,22 +324,25 @@ var _ = Describe("Session", func() {
 	})
 
 	Context("when wrapping out and err", func() {
+		var outWriterBuffer, errWriterBuffer *Buffer
 		BeforeEach(func() {
-			outWriter = NewBuffer()
-			errWriter = NewBuffer()
+			outWriterBuffer = NewBuffer()
+			outWriter = outWriterBuffer
+			errWriterBuffer = NewBuffer()
+			errWriter = errWriterBuffer
 		})
 
 		It("should route to both the provided writers and the gbytes buffers", func() {
 			Eventually(session.Out).Should(Say("We've done the impossible, and that makes us mighty"))
 			Eventually(session.Err).Should(Say("Ah, curse your sudden but inevitable betrayal!"))
 
-			Ω(outWriter.Contents()).Should(ContainSubstring("We've done the impossible, and that makes us mighty"))
-			Ω(errWriter.Contents()).Should(ContainSubstring("Ah, curse your sudden but inevitable betrayal!"))
+			Ω(outWriterBuffer.Contents()).Should(ContainSubstring("We've done the impossible, and that makes us mighty"))
+			Ω(errWriterBuffer.Contents()).Should(ContainSubstring("Ah, curse your sudden but inevitable betrayal!"))
 
 			Eventually(session).Should(Exit())
 
-			Ω(outWriter.Contents()).Should(Equal(session.Out.Contents()))
-			Ω(errWriter.Contents()).Should(Equal(session.Err.Contents()))
+			Ω(outWriterBuffer.Contents()).Should(Equal(session.Out.Contents()))
+			Ω(errWriterBuffer.Contents()).Should(Equal(session.Err.Contents()))
 		})
 	})
 

